File: app/src/main/java/com/ivancea/MTGRules/MainActivity.java
package com.ivancea.MTGRules;

import android.app.AlertDialog;
import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.database.MatrixCursor;
import android.os.Bundle;
import android.speech.tts.TextToSpeech;
import android.view.Menu;
import android.widget.CursorAdapter;
import android.widget.SearchView;
import android.widget.SimpleCursorAdapter;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.ivancea.MTGRules.constants.Actions;
import com.ivancea.MTGRules.constants.Events;
import com.ivancea.MTGRules.model.HistoryItem;
import com.ivancea.MTGRules.model.Rule;
import com.ivancea.MTGRules.model.RulesSource;
import com.ivancea.MTGRules.services.PermissionsRequesterService;
import com.ivancea.MTGRules.services.RulesComparisonService;
import com.ivancea.MTGRules.services.RulesService;
import com.ivancea.MTGRules.services.StorageService;
import com.ivancea.MTGRules.ui.main.AboutFragment;
import com.ivancea.MTGRules.ui.main.MainViewModel;
import com.ivancea.MTGRules.utils.IntentSender;
import com.ivancea.MTGRules.utils.RuleUtils;
import com.ivancea.MTGRules.utils.RulesSearchUtils;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class MainActivity extends AppCompatActivity {

	@Inject
	StorageService storageService;

	@Inject
	RulesService rulesService;

	@Inject
	RulesComparisonService rulesComparisonService;

	@Inject
	PermissionsRequesterService permissionsRequesterService;

	private FirebaseAnalytics mFirebaseAnalytics;

	private MainViewModel viewModel;

	private Boolean ttsOk = null;
	private TextToSpeech tts;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		// Set variables
		mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);
		viewModel = new ViewModelProvider(this).get(MainViewModel.class);
		tts = new TextToSpeech(this, status -> {
			if (status == TextToSpeech.SUCCESS) {
				ttsOk = true;
				tts.setLanguage(Locale.ENGLISH);
			} else {
				ttsOk = false;
			}
		});

		// Save the last rules source
		LocalDate currentRulesSource = rulesService.getLatestRulesSource().getDate();
		storageService.setLastRulesSource(currentRulesSource);

		// Set theme
		boolean useLightTheme = storageService.getUseLightTheme();
		setTheme(useLightTheme);

		// Load ViewModel
		boolean showSymbols = storageService.getShowSymbols();
		viewModel.getShowSymbols().setValue(showSymbols);

		if (viewModel.getCurrentRules().getValue().isEmpty()) {
			RulesSource rulesSource = rulesService.getLatestRulesSource();

			useRules(rulesSource);
		}

		// Inflate layout and fragment
		setContentView(R.layout.main_activity);

		permissionsRequesterService.requestNotifications();

		// Configure action bar
		ActionBar actionBar = getSupportActionBar();
		if (actionBar != null) {
			actionBar.setLogo(R.drawable.ic_launcher_foreground);
			actionBar.setTitle(R.string.app_name);
			viewModel.getActionbarSubtitle().observe(this, actionBar::setSubtitle);
		}


		// Handle initial intent
		handleIntent(getIntent());
	}

	@Override
	protected void onStop() {
		if (tts != null) {
			tts.stop();
		}
		super.onStop();
	}

	@Override
	protected void onDestroy() {
		if (tts != null) {
			tts.shutdown();
		}
		super.onDestroy();
	}

	@Override
	public void onNewIntent(Intent intent) {
		super.onNewIntent(intent);
		handleIntent(intent);
	}

	private void handleIntent(Intent intent) {
		if (intent == null || intent.getAction() == null) {
			return;
		}

		boolean addToHistory = !intent.getBooleanExtra(Actions.IGNORE_HISTORY, false);

		switch (intent.getAction()) {
			case Intent.ACTION_SEARCH -> {
				String searchString = intent.getStringExtra(SearchManager.QUERY);
				String rootRule = intent.getStringExtra(Actions.ROOT_RULE);
				searchRules(searchString, rootRule);

				if (addToHistory) {
					pushHistoryItem(new HistoryItem(HistoryItem.Type.Search, new String[] { searchString, rootRule }));
				}

				logEvent(Events.SEARCH_RULES);
			}

			case Actions.ACTION_READ -> {
				if (ttsOk == null) {
					Toast.makeText(this, R.string.toast_tts_not_ready, Toast.LENGTH_SHORT).show();
				} else if (!ttsOk) {
					Toast.makeText(this, R.string.toast_tts_not_working, Toast.LENGTH_SHORT).show();
				} else {
					tts.speak(intent.getStringExtra(Actions.DATA), TextToSpeech.QUEUE_FLUSH, null, "rules");
					logEvent(Events.READ_RULE);
				}
			}

			case Actions.ACTION_NAVIGATE_RULE -> {
				String title = intent.getStringExtra(Actions.DATA);

				if (title.isEmpty()) {
					viewModel.getVisibleRules().setValue(viewModel.getCurrentRules().getValue());
					viewModel.getSelectedRuleTitle().setValue(null);
					viewModel.getSearchText().setValue(null);

					if (addToHistory) {
						pushHistoryItem(new HistoryItem(HistoryItem.Type.Rule, title));
					}
				} else {
					Rule existingRule = viewModel.getVisibleRules().getValue().stream()
						.filter(r -> r.getTitle().equals(title))
						.findAny().orElse(null);

					if (existingRule == null || !existingRule.getSubRules().isEmpty() || !isLastHistoryItemNavigation()) {
						List<Rule> rules = RuleUtils.getRuleAndParents(
							viewModel.getCurrentRules().getValue(),
							title
						).collect(Collectors.toCollection(ArrayList::new));

						if (!rules.isEmpty()) {
							Rule rule = rules.get(rules.size() - 1);
							if (rule.getSubRules().isEmpty()) {
								rules.remove(rules.size() - 1);
							}

							rules.addAll(rules.get(rules.size() - 1).getSubRules());

							viewModel.getVisibleRules().setValue(rules);

							if (addToHistory) {
								pushHistoryItem(new HistoryItem(HistoryItem.Type.Rule, title));
							}
						}
					}

					viewModel.getSelectedRuleTitle().setValue(title);
					viewModel.getSearchText().setValue(null);
				}
			}

			case Actions.ACTION_RANDOM_RULE -> {
				List<Rule> rules = viewModel.getCurrentRules().getValue();

				List<Rule> allRules = rules.stream()
					.flatMap(RuleUtils::flatten)
					.filter(r -> r.getSubRules().isEmpty())
					.collect(Collectors.toList());

				Random random = new Random();

				int seed = intent.getIntExtra(Actions.DATA, random.nextInt(Integer.MAX_VALUE));
				random.setSeed(seed);

				int rulesToSkip = random.nextInt(allRules.size());

				allRules.stream()
					.skip(rulesToSkip)
					.findFirst()
					.ifPresent(rule -> {
						List<Rule> newVisibleRules = RuleUtils.getRuleAndParents(
							viewModel.getCurrentRules().getValue(),
							rule.getTitle()
						).collect(Collectors.toList());

						viewModel.getVisibleRules().setValue(newVisibleRules);
						viewModel.getSelectedRuleTitle().setValue(null);
						viewModel.getSearchText().setValue(null);

						if (addToHistory) {
							pushHistoryItem(new HistoryItem(HistoryItem.Type.Random, seed));
						}
					});

				logEvent(Events.RANDOM_RULE);
			}

			case Actions.ACTION_CHANGE_THEME -> {
				boolean newUseLightTheme = !storageService.getUseLightTheme();

				setTheme(newUseLightTheme);

				storageService.setUseLightTheme(newUseLightTheme);

				recreate();
			}

			case Actions.ACTION_TOGGLE_SYMBOLS -> {
				boolean newShowSymbols = !storageService.getShowSymbols();

				viewModel.getShowSymbols().setValue(newShowSymbols);

				storageService.setShowSymbols(newShowSymbols);
			}
		}
	}

	private void pushHistoryItem(HistoryItem historyItem) {
		ArrayList<HistoryItem> newHistory = new ArrayList<>(viewModel.getHistory().getValue());
		newHistory.add(historyItem);
		viewModel.getHistory().setValue(newHistory);
	}

	private boolean popHistoryItem() {
		ArrayList<HistoryItem> newHistory = new ArrayList<>(viewModel.getHistory().getValue());

		if (newHistory.isEmpty()) {
			return false;
		}

		newHistory.remove(newHistory.size() - 1);

		while (!newHistory.isEmpty() && newHistory.get(newHistory.size() - 1).getType().equals(HistoryItem.Type.Ignored)) {
			newHistory.remove(newHistory.size() - 1);
		}

		if (newHistory.isEmpty()) {
			return false;
		}

		HistoryItem historyItem = newHistory.get(newHistory.size() - 1);
		viewModel.getHistory().setValue(newHistory);

		switch (historyItem.getType()) {
			case Rule -> {
				IntentSender.openRule(this, (String) historyItem.getValue(), true);
			}
			case Search -> {
				String[] values = (String[]) historyItem.getValue();
				String searchText = values[0];
				String rootRule = values.length == 2 ? values[1] : null;
				IntentSender.openSearch(this, searchText, rootRule, true);
			}
			case Random -> {
				IntentSender.openRandomRule(this, (int) historyItem.getValue(), true);
			}
		}

		return true;
	}

	private boolean isLastHistoryItemNavigation() {
		List<HistoryItem> history = viewModel.getHistory().getValue();
		return !history.isEmpty() &&
			history.get(history.size() - 1).getType().equals(HistoryItem.Type.Rule);
	}

	private void logEvent(String event) {
		mFirebaseAnalytics.logEvent(event, null);
	}

	private void setTheme(boolean useLightTheme) {
		setTheme(useLightTheme ? R.style.LightTheme : R.style.DarkTheme);
	}

	private void searchRules(String searchText, @Nullable String rootRule) {
		List<Rule> rules = viewModel.getCurrentRules().getValue();
		List<Rule> rulesToSearch = rootRule == null ? rules : RuleUtils.getRuleAndSubsections(rules, rootRule);

		List<Rule> filteredRules = RulesSearchUtils.search(searchText, rulesToSearch);

		viewModel.getVisibleRules().setValue(filteredRules);
		viewModel.getSelectedRuleTitle().setValue(null);
		viewModel.getSearchText().setValue(searchText);
	}

	private void useRules(RulesSource rulesSource) {
		List<Rule> rules = rulesService.loadRules(rulesSource);

		viewModel.getCurrentRules().setValue(rules);
		viewModel.getVisibleRules().setValue(rules);
		viewModel.getSelectedRuleTitle().setValue(null);
		viewModel.getSearchText().setValue(null);
		viewModel.getHistory().setValue(Collections.singletonList(new HistoryItem(HistoryItem.Type.Rule, "")));

		viewModel.getActionbarSubtitle()
			.setValue(getString(R.string.action_bar_rules) + ": " + rulesSource.getDate().format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)));
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.main_menu, menu);

		SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
		SearchView searchView = (SearchView) menu.findItem(R.id.search).getActionView();

		searchView.setQueryHint(getString(R.string.search_hint));
		searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));

		String[] columnNames = { SearchManager.SUGGEST_COLUMN_TEXT_1 };
		int[] viewIds = { android.R.id.text1 };
		CursorAdapter adapter = new SimpleCursorAdapter(
			this,
			android.R.layout.simple_list_item_1,
			null,
			columnNames,
			viewIds,
			0
		);
		searchView.setSuggestionsAdapter(adapter);

		searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
			@Override
			public boolean onQueryTextSubmit(String query) {
				return false;
			}

			@Override
			public boolean onQueryTextChange(String newText) {
				String[] columns = new String[] { "_id", SearchManager.SUGGEST_COLUMN_TEXT_1 };

				MatrixCursor cursor = new MatrixCursor(columns);

				String newTextUpperCase = newText.toUpperCase();

				List<Rule> rules = viewModel.getCurrentRules().getValue();

				if (!newText.isEmpty()) {
					rules.get(rules.size() - 1).getSubRules().stream()
						.map(Rule::getTitle)
						.filter(r -> r.toUpperCase().contains(newTextUpperCase))
						.forEach(r -> cursor.newRow().add(r.hashCode()).add(r));
				}

				adapter.changeCursor(cursor);
				adapter.notifyDataSetChanged();

				return true;
			}
		});

		searchView.setOnSuggestionListener(new SearchView.OnSuggestionListener() {
			@Override
			public boolean onSuggestionSelect(int position) {
				return false;
			}

			@Override
			public boolean onSuggestionClick(int position) {
				adapter.getCursor().moveToPosition(position);
				String title = adapter.getCursor().getString(1);

				searchView.clearFocus();

				IntentSender.openRule(MainActivity.this, title, false);

				return true;
			}
		});

		menu.findItem(R.id.home).setOnMenuItemClickListener(view -> {
			IntentSender.openRule(this, "", false);
			return true;
		});

		menu.findItem(R.id.randomRule).setOnMenuItemClickListener(view -> {
			IntentSender.openRandomRule(this, null, false);

			return true;
		});

		menu.findItem(R.id.changeTheme).setOnMenuItemClickListener(view -> {
			IntentSender.changeTheme(this);

			return true;
		});

		menu.findItem(R.id.toggleSymbols).setOnMenuItemClickListener(view -> {
			IntentSender.toggleSymbols(this);

			return true;
		});

		menu.findItem(R.id.compareRules).setOnMenuItemClickListener(view -> {
			List<String> formattedRulesSources = rulesService.getRulesSources().stream()
				.map(rulesSource -> rulesSource.getDate().format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)))
				.collect(Collectors.toCollection(ArrayList::new));

			Collections.reverse(formattedRulesSources);

			showPicker(R.string.dialog_select_source_rules, formattedRulesSources)
				.thenAccept(selectedSourceIndex ->
					showPicker(R.string.dialog_select_target_rules, formattedRulesSources)
						.thenAccept(selectedTargetIndex -> {
							RulesSource sourceRulesSource = rulesService.getRulesSources().get(rulesService.getRulesSources().size() - selectedSourceIndex - 1);
							RulesSource targetRulesSource = rulesService.getRulesSources().get(rulesService.getRulesSources().size() - selectedTargetIndex - 1);

							List<Rule> sourceRules = rulesService.loadRules(sourceRulesSource);
							List<Rule> targetRules = rulesService.loadRules(targetRulesSource);

							List<Rule> comparedRules = rulesComparisonService.compareRules(sourceRules, targetRules);

							viewModel.getVisibleRules().setValue(comparedRules);
							viewModel.getSelectedRuleTitle().setValue(null);
							viewModel.getSearchText().setValue(null);

							pushHistoryItem(new HistoryItem(HistoryItem.Type.Ignored, null));

							logEvent(Events.COMPARE_RULES);
						})
				);

			return true;
		});

		menu.findItem(R.id.changeRules).setOnMenuItemClickListener(view -> {
			List<String> formattedRulesSources = rulesService.getRulesSources().stream()
				.map(rulesSource -> rulesSource.getDate().format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)))
				.collect(Collectors.toCollection(ArrayList::new));

			Collections.reverse(formattedRulesSources);

			showPicker(R.string.dialog_select_rules, formattedRulesSources)
				.thenAccept(selectedIndex -> {
					RulesSource rulesSource = rulesService.getRulesSources().get(rulesService.getRulesSources().size() - selectedIndex - 1);

					useRules(rulesSource);

					logEvent(Events.CHANGE_RULES);
				});

			return true;
		});

		menu.findItem(R.id.about).setOnMenuItemClickListener(view -> {
			new AboutFragment().show(getSupportFragmentManager(), null);

			return true;
		});

		return super.onCreateOptionsMenu(menu);
	}

	private CompletableFuture<Integer> showPicker(int titleId, List<String> pickElement) {
		CompletableFuture<Integer> future = new CompletableFuture<>();

		new AlertDialog.Builder(this)
			.setTitle(titleId)
			.setNegativeButton(R.string.dialog_cancel, (dialog, which) -> {
			})
			.setItems(pickElement.toArray(new String[0]), (dialog, which) ->
				future.complete(which)
			)
			.show();

		return future;
	}

	@Override
	public void onBackPressed() {
		if (!popHistoryItem()) {
			super.onBackPressed();
		}
	}
}

File: app/src/main/java/com/ivancea/MTGRules/MainActivity.kt
package com.ivancea.MTGRules

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.OnBackPressedCallback
import androidx.activity.compose.setContent
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.lifecycle.ViewModelProvider
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.ktx.Firebase
import com.google.firebase.remoteconfig.ktx.remoteConfig
import com.google.firebase.remoteconfig.ktx.remoteConfigSettings
import com.ivancea.MTGRules.constants.Actions
import com.ivancea.MTGRules.constants.Events
import com.ivancea.MTGRules.constants.FirebaseConfig
import com.ivancea.MTGRules.model.HistoryItem
import com.ivancea.MTGRules.model.Rule
import com.ivancea.MTGRules.model.VisibleRules
import com.ivancea.MTGRules.presentation.MainViewModel
import com.ivancea.MTGRules.presentation.main.components.main.MainComponent
import com.ivancea.MTGRules.services.ChangelogService
import com.ivancea.MTGRules.services.PermissionsRequesterService
import com.ivancea.MTGRules.services.RulesComparisonService
import com.ivancea.MTGRules.services.RulesService
import com.ivancea.MTGRules.services.StorageService
import com.ivancea.MTGRules.utils.IntentSender
import com.ivancea.MTGRules.utils.RuleUtils
import com.ivancea.MTGRules.utils.RuleUtils.getRuleAndParents
import com.ivancea.MTGRules.utils.RuleUtils.getRuleAndSubsections
import com.ivancea.MTGRules.utils.RulesSearchUtils.search
import dagger.hilt.android.AndroidEntryPoint
import java.time.LocalDate
import java.util.Locale
import java.util.Random
import java.util.stream.Collectors
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : ComponentActivity() {
    @JvmField
    @Inject
    var storageService: StorageService? = null

    @JvmField
    @Inject
    var rulesService: RulesService? = null

    @JvmField
    @Inject
    var rulesComparisonService: RulesComparisonService? = null

    @JvmField
    @Inject
    var changelogService: ChangelogService? = null

    @JvmField
    @Inject
    var permissionsRequesterService: PermissionsRequesterService? = null

    private var viewModel: MainViewModel? = null
    private var ttsOk: Boolean? = null
    private var tts: TextToSpeech? = null

    private val onBackPressedCallback: OnBackPressedCallback = object : OnBackPressedCallback(false) {
        override fun handleOnBackPressed() {
            if (!popHistoryItem()) {
                finish()
            }
        }
    }

    @OptIn(ExperimentalFoundationApi::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProvider(this).get(MainViewModel::class.java)

        configureFirebaseRemoteConfig()

        tts = TextToSpeech(this) { status: Int ->
            if (status == TextToSpeech.SUCCESS) {
                ttsOk = true
                tts!!.language = Locale.ENGLISH
            } else {
                ttsOk = false
            }
        }

        // Save the last rules source
        val currentRulesSource = rulesService!!.latestRulesSource.date
        storageService!!.lastRulesSource = currentRulesSource

        // Set theme
        val useLightTheme = storageService!!.useLightTheme
        setTheme(useLightTheme)

        // Load ViewModel
        val showSymbols = storageService!!.showSymbols
        viewModel!!.showSymbols.value = showSymbols
        val showAds = storageService!!.showAds
        viewModel!!.showAds.value = showAds

        if (viewModel!!.currentRules.value.isEmpty()) {
            val rulesSource = rulesService!!.latestRulesSource
            viewModel!!.useRules(rulesSource)
        }

        setContent {
            MainComponent()
        }

        permissionsRequesterService!!.requestNotifications()

        changelogService!!.notifyChangelog()

        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)

        // Handle initial intent
        handleIntent(intent)
    }

    private fun configureFirebaseRemoteConfig() {
        Firebase.remoteConfig.setConfigSettingsAsync(
            remoteConfigSettings {
                minimumFetchIntervalInSeconds = 3600
            }
        )

        Firebase.remoteConfig.fetchAndActivate().addOnCompleteListener(this) { task ->
            if (task.isSuccessful) {
                viewModel!!.showAds.value = storageService!!.showAds
                viewModel!!.bannerAdUnitId.value = FirebaseConfig.getBannerAdUnitId()
            }

            viewModel!!.configLoaded.value = true
        }
    }

    override fun onStop() {
        if (tts != null) {
            tts!!.stop()
        }
        super.onStop()
    }

    override fun onDestroy() {
        if (tts != null) {
            tts!!.shutdown()
        }
        super.onDestroy()
    }

    public override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent?) {
        if (intent == null || intent.action == null) {
            return
        }
        val addToHistory = !intent.getBooleanExtra(Actions.IGNORE_HISTORY, false)
        when (intent.action) {
            Actions.ACTION_LOAD_RULES_SOURCE -> {
                val rulesSourceDateString = intent.getStringExtra(Actions.DATA)!!
                val rulesSourceDate = LocalDate.parse(rulesSourceDateString)

                val rulesSource = rulesService!!.rulesSources.find { it.date == rulesSourceDate }

                viewModel!!.useRules(rulesSource!!)
                viewModel!!.logEvent(Events.CHANGE_RULES)
            }

            Intent.ACTION_SEARCH -> {
                val searchString = intent.getStringExtra(SearchManager.QUERY)
                val rootRule = intent.getStringExtra(Actions.ROOT_RULE)
                searchRules(searchString, rootRule)
                if (addToHistory) {
                    pushHistoryItem(
                        HistoryItem(
                            HistoryItem.Type.Search,
                            arrayOf(searchString, rootRule)
                        )
                    )
                }
                viewModel!!.logEvent(Events.SEARCH_RULES)
            }

            Actions.ACTION_READ -> {
                if (ttsOk == null) {
                    Toast.makeText(this, R.string.toast_tts_not_ready, Toast.LENGTH_SHORT).show()
                } else if (!ttsOk!!) {
                    Toast.makeText(this, R.string.toast_tts_not_working, Toast.LENGTH_SHORT).show()
                } else {
                    tts!!.speak(
                        intent.getStringExtra(Actions.DATA),
                        TextToSpeech.QUEUE_FLUSH,
                        null,
                        "rules"
                    )
                    viewModel!!.logEvent(Events.READ_RULE)
                }
            }

            Actions.ACTION_NAVIGATE_RULE -> {
                val title = intent.getStringExtra(Actions.DATA)
                if (title!!.isEmpty()) {
                    viewModel!!.visibleRules.value = VisibleRules.Rules(viewModel!!.currentRules.value)
                    viewModel!!.selectedRuleTitle.value = null
                    viewModel!!.searchText.value = null
                    if (addToHistory) {
                        pushHistoryItem(HistoryItem(HistoryItem.Type.Rule, title))
                    }
                } else if (viewModel!!.selectedRuleTitle.value != title) {
                    val rules = getRuleAndParents(
                        viewModel!!.currentRules.value,
                        title
                    ).collect(Collectors.toCollection { ArrayList() })
                    if (rules.isNotEmpty()) {
                        val rule = rules[rules.size - 1]
                        if (rule.subRules.isEmpty()) {
                            rules.removeAt(rules.size - 1)
                        }
                        rules.addAll(rules[rules.size - 1].subRules)
                        viewModel!!.visibleRules.value = VisibleRules.Rules(rules)
                        if (addToHistory) {
                            pushHistoryItem(
                                HistoryItem(
                                    HistoryItem.Type.Rule,
                                    title
                                )
                            )
                        }
                    }
                    viewModel!!.selectedRuleTitle.value = title
                    viewModel!!.searchText.value = null
                }
            }

            Actions.ACTION_RANDOM_RULE -> {
                val rules = viewModel!!.currentRules.value
                val allRules = rules.stream()
                    .flatMap { obj: Rule -> RuleUtils.flatten(obj) }
                    .filter { r: Rule -> r.subRules.isEmpty() }
                    .collect(Collectors.toList())
                val random = Random()
                val seed = intent.getIntExtra(Actions.DATA, random.nextInt(Int.MAX_VALUE))
                random.setSeed(seed.toLong())
                val rulesToSkip = random.nextInt(allRules.size)
                allRules.stream()
                    .skip(rulesToSkip.toLong())
                    .findFirst()
                    .ifPresent { (title): Rule ->
                        val newVisibleRules = getRuleAndParents(
                            viewModel!!.currentRules.value,
                            title
                        ).collect(Collectors.toList())
                        viewModel!!.visibleRules.value = VisibleRules.Rules(newVisibleRules)
                        viewModel!!.selectedRuleTitle.value = null
                        viewModel!!.searchText.value = null
                        if (addToHistory) {
                            pushHistoryItem(HistoryItem(HistoryItem.Type.Random, seed))
                        }
                    }
                viewModel!!.logEvent(Events.RANDOM_RULE)
            }

            Actions.ACTION_CHANGE_THEME -> {
                val newUseLightTheme = !storageService!!.useLightTheme
                setTheme(newUseLightTheme)
                storageService!!.useLightTheme = newUseLightTheme
                recreate()
            }

            Actions.ACTION_TOGGLE_SYMBOLS -> {
                val newShowSymbols = !storageService!!.showSymbols
                viewModel!!.showSymbols.value = newShowSymbols
                storageService!!.showSymbols = newShowSymbols
            }

            Actions.ACTION_TOGGLE_ADS -> {
                val newShowAds = !storageService!!.showAds
                viewModel!!.showAds.value = newShowAds
                storageService!!.showAds = newShowAds

                val bundle = Bundle()
                bundle.putBoolean(FirebaseAnalytics.Param.VALUE, newShowAds)
                viewModel!!.logEvent(Events.TOGGLE_ADS, bundle)
            }
        }
    }

    private fun pushHistoryItem(historyItem: HistoryItem) {
        viewModel!!.pushHistoryItem(historyItem)

        onBackPressedCallback.isEnabled = isLastHistoryItemNavigation;
    }

    private fun popHistoryItem(): Boolean {
        if (viewModel!!.history.value.isEmpty()) {
            return false
        }

        val newHistory = java.util.ArrayList(
            viewModel!!.history.value
        )
        newHistory.removeAt(newHistory.size - 1)
        while (newHistory.isNotEmpty() && newHistory[newHistory.size - 1].type == HistoryItem.Type.Ignored) {
            newHistory.removeAt(newHistory.size - 1)
        }
        if (newHistory.isEmpty()) {
            return false
        }
        val (type, value) = newHistory[newHistory.size - 1]
        viewModel!!.history.value = newHistory
        when (type) {
            HistoryItem.Type.Rule -> {
                IntentSender.openRule(this, value as String, true)
            }

            HistoryItem.Type.Search -> {
                val values = value as Array<String>
                val searchText = values[0]
                val rootRule = if (values.size == 2) values[1] else null
                IntentSender.openSearch(this, searchText, rootRule, true)
            }

            HistoryItem.Type.Random -> {
                IntentSender.openRandomRule(this, value as Int, true)
            }

            HistoryItem.Type.Ignored -> {
            }
        }
        return true
    }

    /**
     * Checks if there is no more than 1 navigation item (The current one) in the history.
     */
    private val isLastHistoryItemNavigation: Boolean
        get() {
            val history = viewModel!!.history.value

            var foundEntry = false
            for (entry in history) {
                if (entry.type == HistoryItem.Type.Ignored) {
                    if (foundEntry) {
                        return false
                    }
                    foundEntry = true
                }
            }

            return true
        }

    private fun setTheme(useLightTheme: Boolean) {
        viewModel!!.darkTheme.value = !useLightTheme
    }

    private fun searchRules(searchText: String?, rootRule: String?) {
        val rules = viewModel!!.currentRules.value
        val rulesToSearch = if (rootRule == null) rules else getRuleAndSubsections(rules, rootRule)
        val filteredRules = search(
            searchText!!, rulesToSearch
        )
        viewModel!!.visibleRules.value = VisibleRules.Rules(filteredRules)
        viewModel!!.selectedRuleTitle.value = null
        viewModel!!.searchText.value = searchText
    }
}